### PR TITLE
ACME Default Certificate

### LINKED ISSUE #8123: Allow assuming default host if no SNI indication is given (for Let's Encrypt)
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Following up on https://github.com/traefik/traefik/issues/2829, I have a situation where I'd like to proxy Redis (TCP, non-HTTP) connections over TLS with automatic Let's Encrypt certificates. (I am aware that Redis 6 has built-in TLS, but it doesn't have built-in Let's Encrypt 😄)

Redis clients in general do not send SNI indications, so one needs a HostSNI(`*`) on the router for it to basically do any routing. However, with that enabled, Traefik always serves the built-in default certificate since there's no other indication on the connection. `openssl s_client` connects nicely, and I can talk Redis to the underlying server, but the certificate is `CN = TRAEFIK DEFAULT CERT`, of course.

### Takeaway

It would be handy to be able to tell Traefik to assume a given fallback SNI name if none was sent by the peer, e.g. (c.f. the below example configuration)

```toml
[tcp.routers.redis-tls-router.tls]
fallbackSNI = "some-example-domain.example.com"
```

### Testing configuration

My very simple file-based testing configuration is (with DNS names censored)

#### static configuration

```toml
[log]
level = "DEBUG"

[entryPoints]
web = { address = ":80" }
redis-tls = { address = ":55555" }

[certificatesResolvers.le.acme]
email = "example@example.com"
storage = "acme.json"
caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"

[certificatesResolvers.le.acme.httpChallenge]
entryPoint = "web"

[providers.file]
filename = "/root/traefik.dyn.toml"
```

#### traefik.dyn.toml
```toml
[tcp.routers.redis-tls-router]
entryPoints = [ "redis-tls" ]
rule = "HostSNI(`*`)"
service = "redis"

[tcp.routers.redis-tls-router.tls]
certResolver = "le"

[[tcp.routers.redis-tls-router.tls.domains]]
main = "some-example-domain.example.com"

[tcp.services.redis.loadBalancer]
servers = [ { address = "127.0.0.1:6379" } ]
```

### PR DESCRIPTION
**Pull request by rtribotte**
### What does this PR do?

This PR adds the capacity for a TLS Store to use an ACME provider (like Let's Encrypt) to generate the default certificate. 

### Motivation

Fixes #8123.
Supersedes #8124.

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

Co-authored-by: Ludovic Fernandez <ldez@users.noreply.github.com>
Co-authored-by: Julien Salleyron <julien.salleyron@gmail.com>

### CODE REVIEW

File: docs/content/https/tls.md [hunk 1]
```diff
@@ -159,6 +159,75 @@ data:
...
+
+tls:
+  stores:
+    default:
+      defaultCertResolver: myresolver
+      defaultCertDomain:
+        main: foobar
+        sans:
+          - foobar
+          - foobar
```
Comment (ddtmachado) [inline docs/content/https/tls.md:182]: do we still plan on supporting [multiple certificate stores](https://github.com/traefik/traefik/issues/4756)? ;)
If we don't, in my opinion it does not make sense to have `default` prepended on the config keys as `default` is already the name of the single store available

Comment (rtribotte) [reply docs/content/https/tls.md:182]: This is a good point, thus I think that the removal of the multiple TLS stores definition in the configuration should be handled in another PR, if at some point we want to remove it.

File: docs/content/https/tls.md [hunk 2]
```diff
@@ -159,6 +159,75 @@ data:
 
 If no default certificate is provided, Traefik generates and uses a self-signed certificate.
```
Comment (akx) [inline docs/content/https/tls.md:160]: This sentence seems to be at odds with the new chapter below.

File: docs/content/https/tls.md [hunk 3]
```diff
@@ -159,6 +159,75 @@ data:
...
 
+### ACME Default Certificate
+
+You can configure Traefik to use an ACME provider (like Let's Encrypt) to generate the default certificate.
+The configuration to resolve the default certificate should be defined in a TLS store:
+
+!!! important "Precedence with the `defaultCertificate` option"
+
+    The `defaultCertificate` definition takes precedence on the ACME default certificate configuration.
```
Comment (akx) [inline docs/content/https/tls.md:169]:
```suggestion
    The `defaultCertificate` definition takes precedence over the ACME default certificate configuration.
```

I don't think this is quite clear as it is, though, especially if it's important?

File: docs/content/https/tls.md (hunk 1 above)
Comment (akx) [inline docs/content/https/tls.md:182]: Why a triplicate foobar? (The same stands for the other examples.)

File: docs/content/https/tls.md [hunk 4]
```diff
@@ -159,6 +159,75 @@ data:
...
+```yaml tab="File (YAML)"
+# Dynamic configuration
+
+tls:
+  stores:
+    default:
+      defaultCertResolver: myresolver
+      defaultCertDomain:
+        main: foobar
+        sans:
```
Comment (akx) [inline docs/content/https/tls.md:180]: The definitions for `main` and `sans` aren't explained in this document as far as I can see – they probably should, or they should be linked somewhere that does explain them?

File: pkg/provider/acme/provider.go [hunk 5]
```diff
@@ -434,32 +453,45 @@ func (p *Provider) watchNewDomains(ctx context.Context) {
...
 				}
 
-				for routerName, route := range config.HTTP.Routers {
-					if route.TLS == nil || route.TLS.CertResolver != p.ResolverName {
+				if config.TLS == nil {
+					continue
+				}
+
+				for tlsStoreName, tlsStore := range config.TLS.Stores {
+					// FIXME: warn something or enforce elsewhere ?
```
Comment (akx) [inline pkg/provider/acme/provider.go:461]: Should this FIXME be addressed?

File: pkg/provider/acme/provider.go [hunk 6]
```diff
@@ -468,6 +500,54 @@ func (p *Provider) watchNewDomains(ctx context.Context) {
...
+
+	client, err := p.getClient()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get ACME client %w", err)
+	}
+
+	request := certificate.ObtainRequest{
+		Domains:        domains,
+		Bundle:         true,
+		MustStaple:     oscpMustStaple,
```
Comment (akx) [inline pkg/provider/acme/provider.go:529]: Shouldn't this be`ocspMustStaple` (for [OCSP](https://en.wikipedia.org/wiki/Online_Certificate_Status_Protocol))?

File: pkg/provider/acme/provider.go [hunk 7]
```diff
@@ -806,6 +914,28 @@ func (p *Provider) getValidDomains(ctx context.Context, domain types.Domain) ([]
...
+		domain = types.Domain{Main: validDomains[0], SANs: validDomains[1:]}
+	} else {
+		domain = types.Domain{Main: validDomains[0]}
+	}
+
+	for _, cert := range p.certificates {
+		if reflect.DeepEqual(domain, cert.Certificate.Domain) {
+			return true
+		}
+	}
```
Comment (akx) [inline pkg/provider/acme/provider.go:934]: Will this work correctly if e.g. the SANs in a certificate are in a different order than what `p.getValidDomains()` happens to return?

**Review by akx — COMMENTED** [review]
Some code comments & docs clarification requests...

File: pkg/provider/acme/provider.go [hunk 8]
```diff
@@ -365,12 +374,14 @@ func (p *Provider) register(ctx context.Context, client *lego.Client) (*registra
...
+	logger := log.FromContext(ctx)
+
 	if len(domains) == 0 {
-		log.FromContext(ctx).Debug("No domain parsed in provider ACME")
+		logger.Debug("No domain parsed in provider ACME")
 		return
 	}
 
-	log.FromContext(ctx).Debugf("Try to challenge certificate for domain %v found in HostSNI rule", domains)
+	logger.Debugf("Try to challenge certificate for domain %v found in HostSNI rule", domains)
```
Comment (mpl) [inline pkg/provider/acme/provider.go:384]:
```suggestion
	logger.Debugf("Trying to challenge certificate for domain %v found in HostSNI rule", domains)
```

**Review by ldez — APPROVED** [review]
LGTM

File: docs/content/reference/dynamic-configuration/docker-labels.yml [hunk 9]
```diff
@@ -201,3 +201,13 @@
...
+- "traefik.tls.stores.Store0.defaultcertificate.certfile=foobar"
+- "traefik.tls.stores.Store0.defaultcertificate.keyfile=foobar"
+- "traefik.tls.stores.Store0.defaultgeneratedcert.domain.main=foobar"
+- "traefik.tls.stores.Store0.defaultgeneratedcert.domain.sans=foobar, foobar"
+- "traefik.tls.stores.Store0.defaultgeneratedcert.resolver=foobar"
+- "traefik.tls.stores.Store1.defaultcertificate.certfile=foobar"
+- "traefik.tls.stores.Store1.defaultcertificate.keyfile=foobar"
+- "traefik.tls.stores.Store1.defaultgeneratedcert.domain.main=foobar"
+- "traefik.tls.stores.Store1.defaultgeneratedcert.domain.sans=foobar, foobar"
+- "traefik.tls.stores.Store1.defaultgeneratedcert.resolver=foobar"
```
Comment (akx) [inline docs/content/reference/dynamic-configuration/docker-labels.yml:213]: Should these also have the `example.org` / `foo.example.org` / `bar.example.org` nomenclature used in the docs?

File: docs/content/reference/dynamic-configuration/file.toml [hunk 10]
```diff
@@ -463,7 +463,17 @@
       [tls.stores.Store0.defaultCertificate]
         certFile = "foobar"
         keyFile = "foobar"
+      [tls.stores.Store0.defaultGeneratedCert]
+        resolver = "foobar"
+        [tls.stores.Store0.defaultGeneratedCert.domain]
+          main = "foobar"
+          sans = ["foobar", "foobar"]
```
Comment (akx) [inline docs/content/reference/dynamic-configuration/file.toml:470]: Should these also have the `example.org` / `foo.example.org` / `bar.example.org` nomenclature used in the docs?

File: docs/content/reference/dynamic-configuration/file.yaml [hunk 11]
```diff
@@ -518,7 +518,21 @@ tls:
       defaultCertificate:
         certFile: foobar
         keyFile: foobar
+      defaultGeneratedCert:
+        resolver: foobar
+        domain:
+          main: foobar
+          sans:
+            - foobar
+            - foobar
```
Comment (akx) [inline docs/content/reference/dynamic-configuration/file.yaml:527]: Should these also have the `example.org` / `foo.example.org` / `bar.example.org` nomenclature used in the docs?

File: docs/content/reference/dynamic-configuration/file.yaml [hunk 12]
```diff
@@ -518,7 +518,21 @@ tls:
...
```
Comment (akx) [inline docs/content/reference/dynamic-configuration/file.yaml:538]: Should these also have the `example.org` / `foo.example.org` / `bar.example.org` nomenclature used in the docs?

File: docs/content/reference/dynamic-configuration/kv-ref.md [hunk 13]
```diff
@@ -319,8 +319,16 @@
...
+| `traefik/tls/stores/Store0/defaultGeneratedCert/domain/main` | `foobar` |
+| `traefik/tls/stores/Store0/defaultGeneratedCert/domain/sans/0` | `foobar` |
+| `traefik/tls/stores/Store0/defaultGeneratedCert/domain/sans/1` | `foobar` |
+| `traefik/tls/stores/Store0/defaultGeneratedCert/resolver` | `foobar` |
 | `traefik/tls/stores/Store1/defaultCertificate/certFile` | `foobar` |
 | `traefik/tls/stores/Store1/defaultCertificate/keyFile` | `foobar` |
+| `traefik/tls/stores/Store1/defaultGeneratedCert/domain/main` | `foobar` |
+| `traefik/tls/stores/Store1/defaultGeneratedCert/domain/sans/0` | `foobar` |
+| `traefik/tls/stores/Store1/defaultGeneratedCert/domain/sans/1` | `foobar` |
+| `traefik/tls/stores/Store1/defaultGeneratedCert/resolver` | `foobar` |
```
Comment (akx) [inline docs/content/reference/dynamic-configuration/kv-ref.md:331]: Should these also have the `example.org` / `foo.example.org` / `bar.example.org` nomenclature used in the docs?

File: docs/content/reference/dynamic-configuration/marathon-labels.json [hunk 14]
```diff
@@ -201,3 +201,13 @@
...
+"traefik.tls.stores.Store0.defaultcertificate.certfile": "foobar",
+"traefik.tls.stores.Store0.defaultcertificate.keyfile": "foobar",
+"traefik.tls.stores.Store0.defaultgeneratedcert.domain.main": "foobar",
+"traefik.tls.stores.Store0.defaultgeneratedcert.domain.sans": "foobar, foobar",
+"traefik.tls.stores.Store0.defaultgeneratedcert.resolver": "foobar",
+"traefik.tls.stores.Store1.defaultcertificate.certfile": "foobar",
+"traefik.tls.stores.Store1.defaultcertificate.keyfile": "foobar",
+"traefik.tls.stores.Store1.defaultgeneratedcert.domain.main": "foobar",
+"traefik.tls.stores.Store1.defaultgeneratedcert.domain.sans": "foobar, foobar",
+"traefik.tls.stores.Store1.defaultgeneratedcert.resolver": "foobar",
```
Comment (akx) [inline docs/content/reference/dynamic-configuration/marathon-labels.json:213]: Should these also have the `example.org` / `foo.example.org` / `bar.example.org` nomenclature used in the docs?

File: docs/content/https/tls.md [hunk 15]
```diff
@@ -157,6 +157,76 @@ data:
...
+
+```json tab="Marathon"
+labels: {
+  "traefik.tls.stores.default.defaultgeneratedcert.resolver": "myresolver",
+  "traefik.tls.stores.default.defaultgeneratedcert.domain.main": "example.org",
+  "traefik.tls.stores.default.defaultgeneratedcert.domain.sans": "foo.example.org, bar.example.org",
+}
+```
+
 If no default certificate is provided, Traefik generates and uses a self-signed certificate.
```
Comment (akx) [inline docs/content/https/tls.md:230]: This line is at odds with line 160 in the post-image.

```suggestion
```

File: pkg/provider/acme/provider.go [hunk 16]
```diff
@@ -468,16 +548,69 @@ func (p *Provider) watchNewDomains(ctx context.Context) {
...
+		return nil, nil
+	}
+
+	p.resolvingDomains[domainKey] = struct{}{}
+
+	for _, certDomain := range domains {
+		p.resolvingDomains[certDomain] = struct{}{}
+	}
+
+	p.resolvingDomainsMutex.Unlock()
```
Comment (akx) [inline pkg/provider/acme/provider.go:570]: Would it maybe be better to refactor this locked segment into a separate function that could `defer` unlocking?

File: pkg/provider/acme/provider.go [hunk 17]
```diff
@@ -468,16 +548,69 @@ func (p *Provider) watchNewDomains(ctx context.Context) {
...
+		MustStaple:     oscpMustStaple,
+		PreferredChain: p.PreferredChain,
+	}
+
+	cert, err := client.Certificate.Obtain(request)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate a certificate for the domains %v: %w", domains, err)
+	}
+	if cert == nil {
+		return nil, fmt.Errorf("unable to generate a certificate for the domains %v", domains)
```
Comment (akx) [inline pkg/provider/acme/provider.go:593]: Is this... "unknown error"?

File: pkg/provider/acme/provider.go [hunk 18]
```diff
@@ -670,59 +783,79 @@ func (p *Provider) refreshCertificates() {
...
-				logger.Errorf("domains %v renew certificate with no value: %v", cert.Domain.ToStrArray(), cert)
-				continue
-			}
+		logger.Infof("Renewing certificate from LE : %+v", cert.Domain)
+
+		renewedCert, err := client.Certificate.Renew(certificate.Resource{
+			Domain:      cert.Domain.Main,
+			PrivateKey:  cert.Key,
+			Certificate: cert.Certificate.Certificate,
+		}, true, oscpMustStaple, p.PreferredChain)
```
Comment (akx) [inline pkg/provider/acme/provider.go:820]: `oscp` still looks like a typo.

File: pkg/provider/acme/provider.go [hunk 19]
```diff
@@ -761,51 +894,70 @@ func getX509Certificate(ctx context.Context, cert *Certificate) (*x509.Certifica
...
 		}
+
 		cleanDomains = append(cleanDomains, cleanDomain)
 	}
 
 	return cleanDomains, nil
 }
 
+// certExists returns whether a certificate exists for the given domain.
+func (p *Provider) certExists(validDomains []string) bool {
```
Comment (akx) [inline pkg/provider/acme/provider.go:944]: "given domain", yet the parameter is `domains`.

Looking at the impl, this finds a certificate that matches exactly all of the domains given?

File: pkg/tls/certificate.go [hunk 20]
```diff
@@ -185,24 +136,27 @@ func (c *Certificate) AppendCertificate(certs map[string]map[string]*tls.Certifi
 			}
 		}
 	}
+
+	// Guarantees the order to produce a unique cert key.
+	sort.Strings(SANs)
```
Comment (akx) [inline pkg/tls/certificate.go:141]: Considering this array also includes the CN of the cert, `SANs` does not sound like a correct name.

File: pkg/provider/acme/provider.go [hunk 21]
```diff
@@ -761,51 +894,70 @@ func getX509Certificate(ctx context.Context, cert *Certificate) (*x509.Certifica
...
 		}
 	}
 
 	return crt, err
 }
 
-// getValidDomains checks if given domain is allowed to generate a ACME certificate and return it.
-func (p *Provider) getValidDomains(ctx context.Context, domain types.Domain) ([]string, error) {
+// sanitizeDomains checks if given domain is allowed to generate a ACME certificate and return it.
+func (p *Provider) sanitizeDomains(ctx context.Context, domain types.Domain) ([]string, error) {
```
Comment (akx) [inline pkg/provider/acme/provider.go:913]: It's weird that this function does a different thing than the other equally named function in this PR...

File: docs/content/reference/dynamic-configuration/docker-labels.yml (hunk 9 above)
Comment (ldez) [reply docs/content/reference/dynamic-configuration/docker-labels.yml:213]: not the the reference files

File: docs/content/reference/dynamic-configuration/file.toml [hunk 22]
```diff
@@ -463,7 +463,17 @@
...
     [tls.stores.Store1]
       [tls.stores.Store1.defaultCertificate]
         certFile = "foobar"
         keyFile = "foobar"
+      [tls.stores.Store1.defaultGeneratedCert]
+        resolver = "foobar"
+        [tls.stores.Store1.defaultGeneratedCert.domain]
+          main = "foobar"
+          sans = ["foobar", "foobar"]
```
[the remark this answers is not in the document]
Comment (ldez) [reply docs/content/reference/dynamic-configuration/file.toml:479]: not the the reference files

File: docs/content/reference/dynamic-configuration/file.yaml (hunk 12 above)
Comment (ldez) [reply docs/content/reference/dynamic-configuration/file.yaml:538]: not the the reference files

File: docs/content/reference/dynamic-configuration/file.yaml (hunk 11 above)
Comment (ldez) [reply docs/content/reference/dynamic-configuration/file.yaml:527]: not the reference files

File: docs/content/reference/dynamic-configuration/file.toml (hunk 10 above)
Comment (ldez) [reply docs/content/reference/dynamic-configuration/file.toml:470]: not the reference files

File: docs/content/reference/dynamic-configuration/kv-ref.md (hunk 13 above)
Comment (ldez) [reply docs/content/reference/dynamic-configuration/kv-ref.md:331]: not the reference files

File: docs/content/reference/dynamic-configuration/marathon-labels.json (hunk 14 above)
Comment (ldez) [reply docs/content/reference/dynamic-configuration/marathon-labels.json:213]: not the reference files

File: docs/content/reference/dynamic-configuration/docker-labels.yml (hunk 9 above)
Comment (akx) [reply docs/content/reference/dynamic-configuration/docker-labels.yml:213]: Seems weird that they'd refer to creating a cert with the the same CN as two SANs...

File: pkg/provider/acme/provider.go (hunk 16 above)
Comment (ldez) [reply pkg/provider/acme/provider.go:570]: technical refactor are a bad practice.

Comment (akx) [reply pkg/provider/acme/provider.go:570]: One could argue a single +754/-240 line commit with the message "feat: acme default cert" is also bad practice...

Comment (ldez) [reply pkg/provider/acme/provider.go:570]: One could say that I was in a voice chat with the other maintainers and I converted the vocal reviews into commits...

**Review by mmatur — APPROVED** [review]
LGTM

File: pkg/provider/acme/provider.go (hunk 16 above)
Comment (akx) [reply pkg/provider/acme/provider.go:570]: I kind of do know the background, being the author of #8123 and #8124, which are referred to in the PR message. :)

My point is it'll be quite difficult for a future maintainer or contributor to understand why there's been ~700 lines of changes if there's just a single-line commit message.

Comment (ldez) [reply pkg/provider/acme/provider.go:570]: I think you didn't understand what I said.
It does not matter, the feature is now inside Traefik, this is the most important thing.

**Pull request merged**